libqtile/backend/x11/window.py

```diff
@@ -1028,6 +1233,9 @@ def focus(self, warp: bool) -> None:
 
         if self.group:
             self.group.current_window = self
+
+        if getattr(self, "fullscreen", False) and not self.previous_layer[4]:
```

Comment:
Why `getattr(self, "fullscreen", False)` and not `self.fullscreen`? Do we ever have a situation where that attribute does not exist?

Reply:
I think it must have failed at some point for me to write this. Let me test again!

Reply:
It causes a mypy failure `libqtile/backend/x11/window.py:1257: error: "_Window" has no attribute "fullscreen"`

Reply:
I see. I guess what we should do is have `focus()` methods defined on each of the window classes, rather than one big one with if statements checking which class self is. Happy to type:ignore it for now, we can improve things later